pkg/controller/ipam/validate.go

```diff
@@ -107,6 +144,92 @@ func humanReadableIPRanges(ipRanges []crdv1alpha2.SubnetIPRange) string {
 	return fmt.Sprintf("[%s]", strings.Join(strs, ","))
 }
 
+func rangesOverlap(r1, r2 crdv1alpha2.SubnetIPRange) bool {
+	if r1.CIDR == "" {
+		if r2.CIDR == "" {
+			r1start := net.ParseIP(r1.Start).To16()
+			r1end := net.ParseIP(r1.End).To16()
+			r2start := net.ParseIP(r2.Start).To16()
+			r2end := net.ParseIP(r2.End).To16()
+
+			if ipInRange(r1start, r1end, r2start) || ipInRange(r2start, r2end, r1start) {
+				return true
+			}
+		} else {
+			_, cidr2, _ := net.ParseCIDR(r2.CIDR)
+			if cidr2.Contains(net.ParseIP(r1.Start)) || cidr2.Contains(net.ParseIP(r1.End)) {
+				return true
+			}
+			return ipInRange(net.ParseIP(r1.Start).To16(), net.ParseIP(r1.End).To16(), cidr2.IP.To16())
+		}
+	} else {
+		_, cidr1, _ := net.ParseCIDR(r1.CIDR)
+		if r2.CIDR == "" {
+			if cidr1.Contains(net.ParseIP(r2.Start)) || cidr1.Contains(net.ParseIP(r2.End)) {
+				return true
+			}
+			return ipInRange(net.ParseIP(r2.Start).To16(), net.ParseIP(r2.End).To16(), cidr1.IP.To16())
+		} else {
+			_, cidr2, _ := net.ParseCIDR(r2.CIDR)
+			return cidr1.Contains(cidr2.IP) || cidr2.Contains(cidr1.IP)
+		}
+	}
+	return false
+}
+
+func ipInRange(rangeStart, rangeEnd, ip net.IP) bool {
+	// Validate that ip is within start and end of range
+	return bytes.Compare(ip, rangeStart) >= 0 && bytes.Compare(ip, rangeEnd) <= 0
+}
+
+func ipVersion(ip net.IP) int {
+	if ip.To4() != nil {
+		return 4
+	}
+	return 6
+}
+
+func validateIPRange(r crdv1alpha2.SubnetIPRange, poolIPVersion int) (bool, string) {
+	// Validate the integrity of IPs within the IP range
+	gateway := net.ParseIP(r.Gateway)
```

Comment:
Just to save 4 lines of code, can we validate GW IP version here and rely on further checks to fail if IP version differs for other attributes?